zcash_primitives/src/transaction/builder.rs

```diff
@@ -107,8 +120,179 @@ impl Progress {
     }
 }
 
-enum ChangeAddress {
-    SaplingChangeAddress(OutgoingViewingKey, PaymentAddress),
+/// The balance of inputs, outputs, and fees of a transaction under construction.
+pub struct TransactionBalances {
+    total_in: Amount,
+    change_required: Amount,
+    fee_required: Amount,
+}
+
+impl TransactionBalances {
+    /// Constructs a new balance from its constituent parts. Returns an error
+    /// on overflow or underflow of amount arithmetic.
+    pub fn new(
```

Comment:
This appears to be public so that people can implement `FeeStrategy` externally. But currently `FeeStrategy` is sealed because the types in its method can't be named externally. If it is intended to be sealed, then this constructor doesn't need to be public:
```suggestion
    fn new(
```

Reply:
FeeStrategy isn't intended to be sealed, so I'll have to look at what you mean by the types in its method not being able to be named externally.

Reply:
What I mean is that you have a public type in a private module, and it isn't publicly exported elsewhere in a public module. So outside of the crate, no one can `use zcash_primitives::path::to::Type` to import the types needed to implement the trait.

However, I've looked at the crate documentation again, and it seems that the `builder` submodules of the transaction components _are_ exposed publicly (I didn't think they were based on my initial look at the PR), so these types are nameable and `FeeStrategy` isn't sealed AFAICT.